src/net_processing.cpp

```diff
@@ -2363,6 +2363,23 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 }
             }
         }
+        // If we're in IBD, we want outbound peers that will serve us a useful
+        // chain.  Disconnect peers that are behind us or on chains with
+        // insufficient work.
+        if (IsInitialBlockDownload() && nCount != MAX_HEADERS_RESULTS) {
+            // When nCount < MAX_HEADERS_RESULTS, we know we have no more
+            // headers to fetch from this peer.
+            const arith_uint256 &peer_chainwork = nodestate->pindexBestKnownBlock->nChainWork;
+            const arith_uint256 &our_chainwork = std::max(chainActive.Tip()->nChainWork, nMinimumChainWork);
+            if (peer_chainwork < our_chainwork) {
+                // This peer has too little work on their headers chain to help
+                // us sync -- disconnect if using an outbound slot (unless
+                // whitelisted).
+                if (!(pfrom->fInbound || pfrom->fWhitelisted || pfrom->fAddnode)) {
+                    pfrom->fDisconnect = true;
+                }
```

Comment:
I think this needs to be backed off a bit:  Consider-- imagine we start and are fully synced up, but we are still in IsInitialBlockDownload because the network has been slow and so the tip's timestamp is a ways back.  Then we have a peer we get headers from who is a single block behind us.   This would disconnect them, and I think that is probably undesirable.

It's probably a good idea to review all code as if IsInitialBlockDownload() were just return true; and if there would be misbehavior in that case, then the usage is perhaps suboptimal.
An obvious conservative thing to do would be to just use the nMinimumChainWork ... though perhaps too conservative.  It might also be reasonable to use a GetBlockProofEquivalentTime like test, or just the work several blocks back from chainactive tip.   I think if it was conservative enough it could apply at all times, not just IsIninitialBlockDownload().  

Maybe the existence of the second commit means that it would be okay for this one to be less aggressive?

Reply:
I guess with the second commit, we will eventually disconnect some peers if they stay on chains with less work than our tip.  So I think we could just leave this as nMinimumChainWork, to protect against the case that all our peers only have incomplete/bogus chains (which wouldn't be prevented by the second commit).

Reply:
Does it make sense to avoid the push above:
https://github.com/bitcoin/bitcoin/blob/e9d58023aef97d954b7a75da806ee4f22c9760c3/src/net_processing.cpp#L2332
now that the node will disconnect?